lib/db/observed.go

```diff
@@ -0,0 +1,253 @@
+
+func (dl DropListObserved) MarkDevice(device protocol.DeviceID) map[string]bool {
+	folders, ok := dl[device]
+	if !ok {
+		dl[device] = nil
+	}
+	return folders
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID, myID protocol.DeviceID) {
+	for _, dev := range devices {
+		if dev == myID {
+			continue
+		}
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]bool)
+			dl[dev] = folders
+		}
+		folders[folder] = true
+	}
+}
+
+// CleanPendingDevices removes all pending device entries matching a given set of device IDs
+func (db *Lowlevel) CleanPendingDevices(dropList DropListObserved) error {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		l.Warnf("Could not iterate through pending device entries for cleanup: %v", err)
+		return err
+	}
+	defer iter.Release()
+	for iter.Next() {
+		_, err := db.Get(iter.Key())
+		if err != nil {
+			return err
+		}
+		keyDev := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(keyDev) == protocol.DeviceIDLength {
+			// Valid entries are looked up in the drop-list, invalid ones cleaned up
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			_, dropDev := dropList[deviceID]
+			if !dropDev {
+				continue
+			}
+			l.Debugf("Removing marked pending device %v", deviceID)
+		} else {
+			l.Warnln("Invalid pending device entry, deleting from database: %v", keyDev)
+		}
+		if err := db.Delete(iter.Key()); err != nil {
+			l.Warnf("Failed to remove pending device entry: %v", err)
+		}
+	}
+	return nil
+}
+
+// CleanPendingFolders removes all pending folder entries not matching a given set of
+// device IDs, or matching the set of folder IDs associated with those given devices.
+func (db *Lowlevel) CleanPendingFolders(dropList DropListObserved) error {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+	if err != nil {
+		l.Warnf("Could not iterate through pending folder entries for cleanup: %v", err)
+		return err
+	}
+	defer iter.Release()
+	for iter.Next() {
+		_, err := db.Get(iter.Key())
+		if err != nil {
+			return err
+		}
+		if keyDev, ok := db.keyer.DeviceFromPendingFolderKey(iter.Key()); ok {
+			// Valid entries are looked up in the drop-list, invalid ones cleaned up
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			folders, keepDev := dropList[deviceID]
+			// Check the associated set of folders if provided, otherwise drop.
+			if keepDev {
+				if len(folders) == 0 {
+					continue
+				}
+				folderID := db.keyer.FolderFromPendingFolderKey(iter.Key())
+				// Remove only mentioned folder IDs
+				if !folders[string(folderID)] {
+					continue
+				}
+				l.Debugf("Removing marked pending folder %v for %v", string(folderID), deviceID)
+			} else {
+				l.Debugf("Removing pending folder offered by %v", deviceID)
+			}
```

Comment:
This is quite deeply nested. As you already created a type to hold the info about "to be dropped stuff", you could have a `shouldDrop(device, folder)` or maybe `shouldDrop(key deviceFileKey)` method on that type.

Reply:
Not quite that easy because the interpretation is different in these two places, but I'll think about it.

Reply:
In this case I think the nested structure is better.  Using `goto` yields the same challenges as in `PendingFolders()`: having to pre-declare every variable used between goto and the target label.

I also thought about factoring out a predicate function.  But there are several code paths with alternating outcome, and if at some point the decision (`continue` or try to delete) is clear, all further checks should be skipped.  Debug logging stating the reason is very helpful, but would need to happen inside the predicate function.  Not so pretty either.

The nested indentation actually makes clear what parts are not checked if an earlier condition was already sufficient to decide.  So now I only made sure that further checks are always in the last block after an `if` condition for easier mental parsing (pseudocode):

    if this_condition_holds { log(fine) } else {
            do_further_checking...
    }
    fine_do_it_at_last()

Reply:
A part of this discussion comes down to a preference of early return/break/continue/... over using none of those (respectively returns only at the end). And as far as I know where there's neither is "the correct way of doing things". However I think there's value of doing it consistently, and in code I touched in Syncthing it's return early. Obviously consistency must never become a dogma, but personally I don't see a reason here why not to go with something like in the diff below (I just changed code because I think I am faster that way than trying to explain what I mean exactly - no need to use it just because it's there, it's just for demonstration purpose):

```diff
--- a/lib/db/observed.go
+++ b/lib/db/observed.go
@@ -190,28 +190,32 @@ func (db *Lowlevel) CleanPendingFolders(dropList DropListObserved) {
 	}
 	defer iter.Release()
 	for iter.Next() {
-		if keyDev, ok := db.keyer.DeviceFromPendingFolderKey(iter.Key()); !ok {
-			l.Warnf("Invalid pending folder entry, deleting from database: %v", keyDev)
-		} else {
-			// Valid entries are looked up in the drop-list, invalid ones cleaned up
-			deviceID := protocol.DeviceIDFromBytes(keyDev)
-			folders, keepDev := dropList[deviceID]
-			// Check the associated set of folders if provided, otherwise drop.
-			if !keepDev {
-				l.Debugf("Removing pending folder offered by %v", deviceID)
-			} else if len(folders) == 0 {
-				continue
-			} else {
-				folderID := db.keyer.FolderFromPendingFolderKey(iter.Key())
-				// Remove only mentioned folder IDs
-				if _, dropFolder := folders[string(folderID)]; !dropFolder {
-					continue
-				}
-				l.Debugf("Removing marked pending folder %v for %v", string(folderID), deviceID)
+		if db.shouldDeletePendingFolder(iter.Key(), dropList) {
+			if err := db.Delete(iter.Key()); err != nil {
+				l.Warnf("Failed to remove pending folder entry: %v", err)
 			}
 		}
-		if err := db.Delete(iter.Key()); err != nil {
-			l.Warnf("Failed to remove pending folder entry: %v", err)
-		}
 	}
 }
+
+func (db *Lowlevel) shouldDeletePendingFolder(key []byte, dropList DropListObserved) bool {
+	keyDev, ok := db.keyer.DeviceFromPendingFolderKey(key)
+	if !ok {
+		l.Warnf("Invalid pending folder entry, deleting from database: %v", keyDev)
+		return true
+	}
+	// Valid entries are looked up in the drop-list, invalid ones cleaned up
+	deviceID := protocol.DeviceIDFromBytes(keyDev)
+	folders, keepDev := dropList[deviceID]
+	// Check the associated set of folders if provided, otherwise drop.
+	if !keepDev {
+		return true
+	}
+	folderID := db.keyer.FolderFromPendingFolderKey(key)
+	// Remove only mentioned folder IDs
+	if _, dropFolder := folders[string(folderID)]; !dropFolder {
+		return false
+	}
+	l.Debugf("Removing marked pending folder %v for %v", string(folderID), deviceID)
+	return true
+}
```

Reply:
That's a great suggestion, thank you.  I just didn't think of splitting the function, but it's really the most comprehensible and elegant solution.

Btw. I totally agree with the "return early" policy.  It forces you to think about handling the error before returning, which is often forgotten in a deeply nested conditional after all those `}` lines... But no rule without exception, because there are those 40 % code that do not concern error handling ;-)

This could even possibly replace the `goto` logic in the other functions.  I think it's fine that way but if you like I can refactor them as well.